homeassistant/components/template/switch.py

```diff
@@ -219,3 +297,90 @@ async def async_turn_off(self, **kwargs: Any) -> None:
         if self._template is None:
             self._state = False
             self.async_write_ha_state()
+
+
+class TriggerSwitchEntity(TriggerEntity, SwitchEntity, RestoreEntity):
+    """Switch entity based on trigger data."""
+
+    domain = SWITCH_DOMAIN
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        coordinator: TriggerUpdateCoordinator,
+        config: ConfigType,
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(hass, coordinator, config)
+        friendly_name = self._rendered.get(CONF_NAME, DEFAULT_NAME)
+        self._on_script = Script(hass, config[CONF_TURN_ON], friendly_name, DOMAIN)
+        self._off_script = Script(hass, config[CONF_TURN_OFF], friendly_name, DOMAIN)
+
+        self._state: bool | None = None
+        if (tmpl := config.get(CONF_STATE)) is not None and isinstance(
+            tmpl, template.Template
+        ):
+            self._to_render_simple.append(CONF_STATE)
+            self._parse_result.add(CONF_STATE)
+            self._attr_assumed_state = False
+        else:
+            self._attr_assumed_state = True
+
+        self._attr_device_info = async_device_info_to_link_from_device_id(
+            hass,
+            config.get(CONF_DEVICE_ID),
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Restore last state."""
+        await super().async_added_to_hass()
+        if (
+            (last_state := await self.async_get_last_state()) is not None
+            and last_state.state not in (STATE_UNKNOWN, STATE_UNAVAILABLE)
+            # The trigger might have fired already while we waited for stored data,
+            # then we should not restore state
+            and self._state is None
+        ):
+            self._state = last_state.state == STATE_ON
+            self.restore_attributes(last_state)
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle update of the data."""
+        self._process_data()
+
+        if not self.available:
+            self.async_write_ha_state()
+            return
+
+        if not self._attr_assumed_state:
+            raw = self._rendered.get(CONF_STATE)
+            self._state = template.result_as_boolean(raw)
+
+            self.async_set_context(self.coordinator.data["context"])
+            self.async_write_ha_state()
+        elif self._attr_assumed_state and len(self._rendered) > 0:
+            # In case name, icon, or friendly name have a template but
+            # states does not
+            self.async_write_ha_state()
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if device is on."""
+        return self._state
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Fire the on action."""
+        if self._on_script:
+            await self.async_run_script(self._on_script, context=self._context)
```

Comment:
I don't see the need for a new method which takes the script as a parameter. Either `async_run_script` should be defined in this class and grab the required script from the object properties OR just add `this` as one of the variables when calling `self._on_script.async_run(...)`.

Reply:
That new method will be used by all trigger based template entities (light, fan, select, etc) to provide the action scripts with the `this` variable, just like non trigger based template entities.  This function is needed to fix this long standing issue.  It's added here first and it will be added to all other domains at a later date.